Expose pre/post operations and transpositions explicitly, and add documentation.

nical opened the pull request:
This is a big PR, beware that it has a lot of breaking changes in the matrix types.
It implements what I propose in issue #145 and issue #146, namely:

* Remove mul for matrices in favor of explicit pre_mul and post_mul (similar to what we have in gecko).
* Add Row-major and column-major serializers and constructors to matrices to help interfacing with dom matrices and external libs such as skia (the actual layout of the matrix types stays row-major but that's an implementation detail that only matters to the people looking at the memory in a debugger).
* Add a whole bunch of methods that are used a lot in gecko.
* Add a lot of documentation.

This is still a work in progress but it's almost ready, so interested parties can start looking at it give feedback.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/euclid/149)
<!-- Reviewable:end -->

nical commented:
The PR is in a good state now (modulo making the git history pretty once eventual review comments have been addressed). Here is a summary of what's in there:
 * make pre and post multiplication explicit (removing mul in favor of pre_mul/post_mul
 * implement the common pre/post operations (translation, scale, rotation), using ```translated``` instead of ```translate``` in the terminology to avoid immutable methods looking like they are in-place transformations (and let the names available for in-place operations if we ever find out that it helps with code-gen in the future). see ```post_translated```, ```pre_rotated```, etc.
 * add explicit row-major/column-major constructors and serializers (```row_major``` replaces ```new```, ```column_major``` is added, ```to_array``` becomes ```to_row_major_array``` and ```to_column_major_array``` complements it. With this, the internal layout is completely irrelevant unless you are looking at your stack in a debugger.

The above changes make it hard to fall into any trap by habit of using a convention or another.
Additionally, I added methods that are used all over the place in gecko:
 * float to integer conversion for points/sizes/etc. without these people tend to just use casts but truncating decimals towards zero doesn't always make sense from a geometrical point of view. See ```floor```, ```ceil``` and ```round```.
 * Similar methds for rects, see ```round_in``` and ```round_out```. They are used so much in gecko that I am surprised servo doesn't have equivalents already, or perhaps it is outside of euclid.
 * Lots and lots of documentation all over euclid.

And last but not least, I made a lot of code that was relying on Clone, rely on Copy instead. This simplifies things a great deal and I don't think anyone will ever use euclid with scalar types that are not Copy, although that's debatable. A lot of code was already only implemented for Copy types, especially in matrices, so I just generalized it to the rest of the library.

nical commented:
I realize that's a lot for one PR, sorry. We can discuss, the different aspects of it here and I'll remove things if need be.

samlh commented on src/matrix2d.rs:
```diff
@@ -7,83 +7,199 @@
-impl<T: Copy, Src, Dst> TypedMatrix2D<T, Src, Dst> {
-    pub fn to_array(&self) -> [T; 6] {
+    /// Create a matrix specifying its components in column-major order.
+    pub fn colum_major(m11: T, m21: T, m31: T, m12: T, m22: T, m32: T) -> TypedMatrix2D<T, Src, Dst> {
```
column

samlh commented on src/point.rs:
```diff
@@ -177,42 +210,86 @@ impl<T: Copy + Div<T, Output=T>, U1, U2> Div<ScaleFactor<T, U1, U2>> for TypedPo
-        match (NumCast::from(self.x.clone()), NumCast::from(self.y.clone())) {
+    ///
+    /// When casting from floating point to integer coordinates, the decimals are truncated
+    /// as one would expect from a simple cast, but this behavior does not always marke sense
```
make

peterjoel commented:
I'm curious why the change to make various types implement `Copy`. Is that being tracked or discussed in an issue?

peterjoel commented:
Never mind: somehow I thought you had made the matrix types implement Copy. I see now that it's just a constraint on the scalar type parameter. It's hard to imagine how that would be a problem.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/euclid/149#-:-KO_tIUVAKRgALjJGVFA:b-mly7ae)*
<!-- Sent from Reviewable.io -->

peterjoel commented:
Is there any reason not to also change the `Clone` constraint to `Copy` for `Length`?

peterjoel commented on src/point.rs:
```diff
@@ -77,35 +79,43 @@ impl<T: fmt::Display, U> fmt::Display for TypedPoint2D<T, U> {
     #[inline]
     pub fn from_lengths(x: Length<T, U>, y: Length<T, U>) -> TypedPoint2D<T, U> {
-        TypedPoint2D::new(x.get(), y.get())
+        TypedPoint2D::new(x.0, y.0)
```
`x.get()` actually feels cleaner to me.

nical commented:
Since a lot of methods were already implemented on top of Copy scalar types, having the other half of the library implemented on top of (the less constrained) Clone trait was not particularly useful. Basing everything on top of Copy simplified the code quite a bit.

No particular reason for Length's Clone constraint being left to Clone instead of Copy. For vector/matrix types the incentive was to make the methods easier to write and more readable while Length doesn't do anything fancy with its value. I can change it to use Copy as well If you prefer.

peterjoel commented:
It might be a bit neater though to have the same constraints everywhere. But it probably doesn't matter too much either way.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #153) made this pull request unmergeable. Please resolve the merge conflicts.

peterjoel commented:
I tried applying this to servo. The only thing missing is that `TypedMatrix2D` needs `Copy` and `Clone` impls for `T: Copy`.

peterjoel commented:
When this is merged, I have made the relevant changes to Servo, which I'll make follow-on PRs for:
* https://github.com/peterjoel/rust-layers/commit/24833a45ecfe5909ad46012c11d31bad5e114e9c
* https://github.com/peterjoel/webrender/commit/e74f517974e78a03186c9747782336a9bd3908e1
* https://github.com/peterjoel/servo/commit/7f83feefc99284bc89e38de7ad046a2972a36394

peterjoel commented on src/matrix4d.rs:
```diff
@@ -385,41 +466,94 @@ where T: Copy + Clone +
-    /// https://drafts.csswg.org/css-transforms/#funcdef-skew
+    ///
+    /// See https://drafts.csswg.org/css-transforms/#funcdef-skew
     pub fn create_skew(alpha: T, beta: T) -> TypedMatrix4D<T, Src, Dst> {
```
Shouldn't `alpha` and `beta` be `Radians<T>` here?

nical commented:
Hi, I'm just back from holiday, I'll catch up with this asap, probably tomorrow, address comments, and rebase the PR.
@peterjoel thanks a lot for the followup PRs! I can take care of the rest if you want (I don't mean to have you spend your time cleaning after my breaking changes).

peterjoel commented:
@nical No worries. I need to keep my unmerged DOMMatrix branch up to date, so I'm applying those changes anyway.

nical commented:
I just rebased the branch, squashed it into a single commit fixed the skew parameters to use radians explicitly. I also rolled back the part where Point::zero() was renamed into Point::origin() which had landed in this branch accidentally.

nical commented:
r? @pcwalton

pcwalton commented:
@bors-servo: r+

Very nice to see more work being done on euclid :)

bors-servo commented:
:pushpin: Commit fe20f19 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton fe20f19d95c4c8355fa83060c6783e6a90a96686 -->

bors-servo commented:
:zap: Test exempted - [status](https://travis-ci.org/servo/euclid/builds/153047234)
